### PR TITLE
[dynatraceexporter] retry, queueing, and error handling

### PR DESCRIPTION
**Pull request by dyladan**
**Description:**

- Use exporterhelper models to implement queue, retry, and timeout configuration
- Additional error handling on error responses from Dynatrace

**Documentation:** README updated to reflect updates

### CODE REVIEW

File: exporter/dynatraceexporter/factory.go [hunk 1]
```diff
@@ -77,7 +82,10 @@ func createMetricsExporter(
 		cfg,
 		params.Logger,
 		exp.PushMetricsData,
-		exporterhelper.WithQueue(exporterhelper.DefaultQueueSettings()),
-		exporterhelper.WithRetry(exporterhelper.DefaultRetrySettings()),
+		// explicitly disable timeout since we rely on http.Client timeout logic.
```
Comment (bogdandrutu) [inline exporter/dynatraceexporter/factory.go:85]: If this is the case you can simply remove the `TimeoutSettings` from the config. the `timeout` config option will be exposed by the `HttpClientSettings`

Comment (dyladan) [reply exporter/dynatraceexporter/factory.go:85]: OK. I followed the example of the splunk and signalfx exporters so it's possible I misunderstood something.

Comment (dyladan) [reply exporter/dynatraceexporter/factory.go]: Removed in 3ec9e4255e6dd64b88d38fbcf0a01e6708ad05af

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 2]
```diff
@@ -50,8 +51,23 @@ const (
 	maxMetricKeyLen = 250
 )
 
-// PushMetricsData exports the given metrics data
-func (e *exporter) PushMetricsData(ctx context.Context, md pdata.Metrics) (droppedTimeSeries int, err error) {
+func (e *exporter) PushMetricsData(ctx context.Context, md pdata.Metrics) (int, error) {
+	request, dropped := e.serializeMetrics(md)
+
+	if request == "" {
```
Comment (arminru) [inline exporter/dynatraceexporter/metrics_exporter.go:57]: Please add a comment on what this case covers

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 3]
```diff
@@ -94,59 +110,67 @@ func (e *exporter) PushMetricsData(ctx context.Context, md pdata.Metrics) (dropp
...
-			return 0, fmt.Errorf("error processing data: %s", err.Error())
-		}
-	}
-
-	return dropped, nil
+	return output, dropped
 }
 
-func (e *exporter) send(ctx context.Context, message string) error {
+func (e *exporter) send(ctx context.Context, message string) (int, error) {
```
Comment (arminru) [inline exporter/dynatraceexporter/metrics_exporter.go:116]: Please add a comment on what the returned int represents

File: exporter/dynatraceexporter/metrics_exporter.go [hunk 4]
```diff
@@ -94,59 +110,67 @@ func (e *exporter) PushMetricsData(ctx context.Context, md pdata.Metrics) (dropp
...
+			return responseBody.Invalid, nil
 		}
 	}
 
-	if !(resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusAccepted) {
-		return fmt.Errorf("request failed with response code: %d", resp.StatusCode)
+	if resp.StatusCode == 401 || resp.StatusCode == 403 || resp.StatusCode == 413 {
+		// Unauthorized and Unauthenticated errors are permanent
+		// If a payload is too large, resending it will not help
+		return 0, consumererror.Permanent(fmt.Errorf(resp.Status))
```
Comment (arminru) [inline exporter/dynatraceexporter/metrics_exporter.go:162]: Is this permanent per lifetime of the collector/exporter or per metrics batch?

File: exporter/dynatraceexporter/factory.go (hunk 1 above)
Comment (dyladan) [reply exporter/dynatraceexporter/factory.go]: @bogdandrutu this pattern is also visible in the official `otlphttpexporter` https://github.com/open-telemetry/opentelemetry-collector/blob/master/exporter/otlphttpexporter/factory.go#L123

File: exporter/dynatraceexporter/metrics_exporter.go (hunk 4 above)
Comment (dyladan) [reply exporter/dynatraceexporter/metrics_exporter.go:162]: Just the batch. We could shutdown on a lifetime error like 401/403 if you prefer?

Comment (arminru) [reply exporter/dynatraceexporter/metrics_exporter.go:178]: Since they are unrecoverable I'd shut down the exporter.

**Review by bogdandrutu — APPROVED** [review]

**Comment by bogdandrutu** [comment]
@arminru this looks good from the collector framework perspective. I need you to also approve from the dynatrace perspective.

**Review by arminru — APPROVED** [review]

**Pull request merged**